### ISSUE TITLE
Functionality to Pause & Resume Uploads

### ISSUE THREAD

**Issue by BuchholzTim** [pqina/filepond#697]
**Is your feature request related to a problem? Please describe.**
I want to prevent a User to upload the same file multiple times. To do this I'd like to get a hash over the content of a file before it actually starts uploading. 

Based on the hash I could then decide whether to start the upload or cancel it.

**Describe the solution you'd like**
The hashing can already be done by catching the file-object through the `addfilestart` event. Unfortunately I cannot prevent the upload (as far as I can tell), while I'm calculating the hash. Depending on the file size this can take some time, in which part of the file will already be uploaded.

So a method to pause/resume the upload of a file would be perfect.

**Describe alternatives you've considered**
I tried removing every file added, hashing it, then re-adding it manually. This works for single files but (currently) breaks for multiple files dropped at the same time. Also for the end-user this solution can be really weird, as the dropped file disappears for some time, while it's being hashed, only to reappear again.

**Comment by rikschennink** [comment]
Can you not calculate hash in a custom process method? If the hash exists → error() → else → upload()

**Comment by BuchholzTim** [comment]
Thanks for the suggestion. I forgot I could actually write my own custom `process`.
However I'm somewhat confused by the documentation. I am always uploading in chunks, so `process` only gets the ID from the server, which is then used by `patch` to upload the file.

I tried implementing my own `process` function but am kinda stuck, perhaps you could help me out. 

```js
process: (
  fieldName,
  file,
  metadata,
  load,
  error,
  progress,
  abort,
  transfer,
  options,
) => {
  // fieldName is the name of the input field - No direct relevance for us
  // logger.log(`fieldName: ${fieldName}`);

  // Usually Empty - Can be added with Metadata-Plugin
  // logger.log(metadata);

  // Contains Information such as chunkSize, chunkUploads=true, etc.
  // logger.log(options);

  // Load & Progress

  // Needed Parameters of file
  const { name, size } = file;

  // API-Endpoint
  const url = `${this.$config.api_url}/filepond/process`;

  /**
   * Hash File
   */

  // Get Cancel-Token, If user wants to abort 
  const source = this.$axios.CancelToken.source();

  // Check if hash is known - If yes, abort ; else progress
  this.hashFile(file).then((hash) => {
    if (hash.length === 0) {
      error('Hash not computable');
    // else if (Hash is already known)
    } else {
      const axiosOptions = {
        cancelToken: source.token,
        headers: {
          'file-category': this.category,
          'crane-project-id': this.craneProjectId,
          'upload-length': size,
          'upload-name': name,
        },
      };
      this.$axios.$post(url, {}, axiosOptions).then((res) => {
        // Pass UUID from Server to Patch() - This is used when uploading in chunks
        transfer(res);
        // Somehow trigger patch?
      });
    }
  });
  
  // Should expose an abort method so the request can be cancelled
  return {
    abort: () => {
      // This function is entered if the user has tapped the cancel button
      source.cancel('Operation cancelled by user');
      // Let FilePond know the request has been cancelled
      abort();
    },
  };
},
```

As far as I can tell from the documentation `transer` is used to pass the ID returned by `process` to Filepond. 
`progress` should only manipulate the indicated Progress and `load` changes to Filepond-Indicator to green and makes it appear 'finished'.

Is there some way to initiate the upload through the `patch`-function from inside my custom `process` after I received to ID from the server or do I get something horribly wrong here?

**Comment by BuchholzTim** [comment]
Okay I think I got it wrong how Filepond handles the Processing of chunked files. I thought there is a `process`-function which literally only handles the first inital contact to the server if the file has to be uploaded in chunks and a seperate `patch`-function, which then uploads the chunks.

I looked into the source-code and if I understand it correctly now filepond uses different `process`-functions depending on whether or not the file has to be uploaded in chunks. If it has to, I think the whole process is handled in `processFileChunked.js`, which exports `processFileChunked`.

So to achieve my goal of hashing each file before actually uploading it I would need to write a custom `process`-function, which would do my stuff (hashing, validating, etc.). Then the function also needs to upload the file in chunks, while updating the progress indicated through `progress` and completing it with the `load`-function.

Is this correct?
